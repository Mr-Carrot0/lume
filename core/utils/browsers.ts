/**
 * Code generated from:
 *
 * ```ts
 * import { getCompatibleVersions } from "npm:baseline-browser-mapping@2.4.4";
 *
 * const results = getCompatibleVersions({
 *   widelyAvailableOnDate: "2025-05-07", // Lume 3.0.0 release date
 * });
 * ```
 *
 * Not run directly to avoid importing this package that requires read access.
 */
export const browsers = {
  chrome: [107, 0],
  chrome_android: [107, 0],
  edge: [107, 0],
  firefox: [104, 0],
  firefox_android: [104, 0],
  safari: [16, 0],
  safari_ios: [16, 0],
} as Record<string, [number, number]>;

export function version(browser: [number, number]): number {
  return (browser[0] << 16) | (browser[1] << 8);
}
export function versionString(browser: [number, number]): string {
  return `${browser[0]}.${browser[1]}`;
}
