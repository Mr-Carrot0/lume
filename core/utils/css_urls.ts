type UrlWalker = (url: string, type: string) => string | Promise<string>;
type Position = [string, string, number, number];
type Index = [string, number];

export async function walkUrls(
  code: string,
  walker: UrlWalker,
): Promise<string> {
  let result = "";
  let index = 0;
  let position = next(code, index);

  while (position) {
    const [type, url, start, end] = position;
    const newUrl = await walker(url, type);
    const quotedUrl = url.startsWith("data:")
      ? `'${newUrl.replaceAll("'", "\\'")}'`
      : `"${newUrl.replaceAll('"', '\\"')}"`;
    result += `${code.slice(index, start)}${quotedUrl}`;
    index = end;
    position = next(code, index);
  }

  return result + code.slice(index);
}

function next(code: string, index = 0): Position | undefined {
  const nextUrl = code.indexOf("url(", index);
  const nextComment = code.indexOf("/*", index);
  const nextImport = code.indexOf("@import ", index);

  let found: Index | undefined;

  if (nextUrl !== -1) {
    found = ["url", nextUrl];
  }
  if (nextComment !== -1 && (!found || nextComment < found[1])) {
    found = ["comment", nextComment];
  }
  if (nextImport !== -1 && (!found || nextImport < found[1])) {
    found = ["import", nextImport];
  }

  if (!found) {
    return;
  }

  // If it's a comment, find the end of it
  if (found[0] === "comment") {
    const end = code.indexOf("*/", found[1]);
    if (end === -1) {
      return;
    }
    return next(code, end + 2);
  }

  if (found[0] === "import") {
    const double = code.indexOf('"', found[1]);
    const single = code.indexOf("'", found[1]);
    const urlFn = code.indexOf("url(", found[1]);

    let foundImport: Index | undefined;

    if (double !== -1) {
      foundImport = ["double", double];
    }
    if (single !== -1 && (!foundImport || single < foundImport[1])) {
      foundImport = ["single", single];
    }

    // It's has a url(), so delegate to the url() logic
    if (urlFn !== -1 && (!foundImport || urlFn < foundImport[1])) {
      return next(code, urlFn);
    }

    if (!foundImport) {
      return;
    }

    const start = foundImport[1];
    const end =
      code.indexOf(foundImport[0] === "double" ? '"' : "'", start + 1) + 1;
    return ["url", code.slice(start + 1, end - 1), start, end];
  }

  if (found[0] === "url") {
    const start = code.indexOf("(", found[1]) + 1;
    let url = code.slice(start).trim();
    const quote = url.startsWith('"') || url.startsWith("'")
      ? url[0]
      : undefined;
    url = quote ? url.slice(1) : url;
    let end = findUrlEnd(url, quote);
    url = url.slice(0, end);
    end += start + (quote ? 2 : 0);
    url = url.replace(/\\'/g, "'").replace(/\\"/g, '"');
    return ["url", url, start, end];
  }
}

function findUrlEnd(code: string, quote?: string): number {
  if (!quote) {
    return code.indexOf(")");
  }

  let end = code.indexOf(quote);

  while (end !== -1 && code[end - 1] === "\\") {
    end = code.indexOf(quote, end + 1);
  }

  return end === -1 ? code.indexOf(")") : end;
}
