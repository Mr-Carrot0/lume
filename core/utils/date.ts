/**
 * Returns the date of the git commit that created or modified the file.
 * Thanks to https://github.com/11ty/eleventy/blob/8dd2a1012de92c5ee1eab7c37e6bf1b36183927e/src/Util/DateGitLastUpdated.js
 */
export function getGitDate(
  type: "created" | "modified",
  file: string,
): Date | undefined {
  const args = type === "created"
    ? ["log", "--diff-filter=A", "--follow", "-1", "--format=%at", "--", file]
    : ["log", "-1", "--format=%at", "--", file];

  const { stdout, success } = new Deno.Command("git", { args }).outputSync();

  if (!success) {
    return;
  }
  const str = new TextDecoder().decode(stdout);

  if (str) {
    return parseDate(parseInt(str) * 1000);
  }
}

/** Parse a string or number (of miliseconds) to UTC Date */
export function parseDate(date: string | number): Date {
  return new Date(getZonedDateTime(date).epochMilliseconds);
}

/** Parse a string or number (of miliseconds) to a zoned datetime */
export function getZonedDateTime(
  date: string | number,
  timezone = "UTC",
): Temporal.ZonedDateTime {
  if (typeof date === "number") {
    return Temporal.Instant.fromEpochMilliseconds(date).toZonedDateTimeISO(
      timezone,
    );
  }

  try {
    return Temporal.Instant.from(date).toZonedDateTimeISO(timezone);
  } catch {
    return Temporal.PlainDateTime.from(date).toZonedDateTime(timezone);
  }
}
