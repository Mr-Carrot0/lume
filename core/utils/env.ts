const envVars = new Map<string, string>();

export function setEnv(name: string, value: string) {
  if (allowedEnvVars()) {
    Deno.env.set(name, value);
  }
  envVars.set(name, value);
}

export function env<T>(name: string): T | undefined {
  const value = allowedEnvVars()
    ? envVars.get(name) ?? Deno.env.get(name)
    : envVars.get(name);

  if (typeof value === "undefined") {
    return undefined;
  }

  switch (value.toLowerCase()) {
    case "true":
    case "on":
    case "1":
      return true as T;

    case "false":
    case "off":
    case "0":
      return false as T;

    default:
      return value as T;
  }
}

let allowed: boolean | undefined;
function allowedEnvVars(): boolean {
  if (allowed === undefined) {
    allowed = Deno.permissions.querySync?.({ name: "env" }).state === "granted";
  }
  return allowed;
}
