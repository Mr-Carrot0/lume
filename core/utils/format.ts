const byteUnit = Intl.NumberFormat("en", {
  notation: "compact",
  style: "unit",
  unit: "byte",
  unitDisplay: "narrow",
});

export function bytes(bytes: number) {
  return byteUnit.format(bytes);
}

const percentageUnit = Intl.NumberFormat("en", {
  style: "percent",
  minimumFractionDigits: 1,
  maximumFractionDigits: 1,
});

export function percentage(initialValue: number, finalValue: number) {
  if (initialValue === finalValue) {
    return "0%";
  }

  const ratio = -(1 - finalValue / initialValue);

  if (ratio < 0) {
    return percentageUnit.format(ratio);
  }

  return "+" + percentageUnit.format(ratio);
}

const secondUnit = Intl.NumberFormat("en", {
  notation: "compact",
  style: "unit",
  unit: "second",
  unitDisplay: "short",
  minimumFractionDigits: 1,
  maximumFractionDigits: 1,
});

const milisecondUnit = Intl.NumberFormat("en", {
  notation: "compact",
  style: "unit",
  unit: "millisecond",
  unitDisplay: "short",
  maximumFractionDigits: 0,
});

export function duration(miliseconds: number) {
  if (miliseconds < 1000) {
    return milisecondUnit.format(miliseconds);
  }

  return secondUnit.format(miliseconds / 1000);
}
