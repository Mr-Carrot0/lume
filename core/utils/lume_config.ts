/** A list of the available plugins not installed by default and sorted */
export const pluginNames = [
  // Order doesn't matter, but should be first
  "attributes",
  "toml",
  "date",
  "code_highlight",
  "decap_cms",
  "fff",
  "eta",
  "extract_date",
  "jsx",
  "json_ld",
  "reading_info",
  "relations",
  "mdx",
  "metas",
  "multilanguage",
  "nav",
  "nunjucks",
  "pagefind",
  "plaintext",
  "prism",
  "pug",
  "remark",
  "robots",
  "sheets",
  "filter_pages",
  "redirects",

  // CSS + JS + source maps
  "esbuild",
  "terser",
  "katex",
  "google_fonts",
  "sass",
  "lightningcss",
  "postcss",
  "tailwindcss",
  "unocss",
  "purgecss",
  "source_maps",

  // Modify URLs
  "base_path",
  "resolve_urls",
  "relative_urls",
  "slugify_urls",
  "modify_urls",
  "check_urls",

  // Images
  "og_images",
  "favicon",
  "svgo",
  "picture",
  "transform_images",

  // Assets in HTML
  "inline",
  "sri",

  // Generate files with URLs
  "feed",
  "sitemap",

  // Final minification and compression
  "minify_html",
  "brotli",
  "gzip",
];

/** Returns the _config file of a site */
export async function getConfigFile(
  path?: string,
  defaultPaths: string[] = ["_config.js", "_config.ts"],
): Promise<string | undefined> {
  if (path) {
    try {
      return await Deno.realPath(path);
    } catch {
      throw new Error(`Config file not found (${path})`);
    }
  }

  for (const path of defaultPaths) {
    try {
      return await Deno.realPath(path);
    } catch {
      // Ignore
    }
  }
}
