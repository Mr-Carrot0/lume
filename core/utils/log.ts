import { env } from "./env.ts";
import {
  bold,
  brightGreen,
  cyan,
  gray,
  italic,
  red,
  strikethrough,
  yellow,
} from "../../deps/colors.ts";

const severity = {
  TRACE: 1,
  DEBUG: 5,
  INFO: 9,
  WARN: 13,
  ERROR: 17,
  FATAL: 21,
};

type LevelName = keyof typeof severity;

// Get the log level from the environment variable LUME_LOGS
const level = env<LevelName>("LUME_LOGS")?.toUpperCase() as
  | LevelName
  | undefined ?? "INFO";

const COLOR_TAG_REG = /<(\w+)>([^<]+)<\/\1>/g;

const logFormats: Record<string, (str: string) => string> = {
  cyan,
  Cyan: (str: string) => bold(cyan(str)),
  red,
  Red: (str: string) => bold(red(str)),
  gray,
  Gray: (str: string) => bold(gray(str)),
  green: brightGreen,
  Green: (str: string) => bold(brightGreen(str)),
  yellow: yellow,
  Yellow: (str: string) => bold(yellow(str)),
  del: (str: string) => strikethrough(gray(str)),
  em: italic,
};

/**
 * This is the default logger. It will output color coded log messages to the
 * console via `console.log()`.
 */
class Logger {
  #level: number;

  constructor(level: LevelName) {
    this.#level = severity[level];
  }

  #log(msg: string, level: number): void {
    if (level >= severity.FATAL) {
      msg = `<Red>FATAL</Red> ${msg}`;
    } else if (level >= severity.ERROR) {
      msg = `<red>ERROR</red> ${msg}`;
    } else if (level >= severity.WARN) {
      msg = `<yellow>WARN</yellow> ${msg}`;
    }

    msg = msg.replaceAll(
      COLOR_TAG_REG,
      (all, name, content) => logFormats[name]?.(content) ?? all,
    );

    if (level >= severity.ERROR) {
      return console.error(msg);
    }

    if (level >= severity.WARN) {
      return console.warn(msg);
    }

    console.log(msg);
  }

  fatal(msg: string): void {
    this.#log(msg, severity.FATAL);
  }

  error(msg: string): void {
    if (this.#level < severity.FATAL) {
      this.#log(msg, severity.ERROR);
    }
  }

  warn(msg: string): void {
    if (this.#level < severity.ERROR) {
      this.#log(msg, severity.WARN);
    }
  }

  info(msg: string): void {
    if (this.#level < severity.WARN) {
      this.#log(msg, severity.INFO);
    }
  }

  debug(msg: string): void {
    if (this.#level < severity.INFO) {
      this.#log(msg, severity.DEBUG);
    }
  }

  trace(msg: string): void {
    if (this.#level < severity.DEBUG) {
      this.#log(msg, severity.TRACE);
    }
  }
}

export const log = new Logger(level);
