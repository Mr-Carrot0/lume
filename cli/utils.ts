import lume from "../mod.ts";
import { log } from "../core/utils/log.ts";
import Site from "../core/site.ts";

/** Create a site instance */
export async function createSite(_config?: URL): Promise<Site> {
  if (!_config) {
    return lume();
  }
  log.info(`Loading config file <gray>${_config}</gray>`);
  const mod = await import(_config.toString());
  const site = mod.default as Site | undefined;
  if (!(site instanceof Site)) {
    log.fatal(
      `[Lume] Missing Site instance! Ensure your config file does export the Site instance as default.`,
    );
    throw new Error("Site instance is not found");
  }
  return site;
}

/** Create a site intance and build it */
export async function buildSite(site: Site): Promise<void> {
  performance.mark("start");
  await site.build();
  performance.mark("end");

  log.info(`üçæ Site built into <gray>${site.options.dest}</gray>`);
  const duration = performance.measure("duration", "start", "end").duration /
    1000;
  const total = site.pages.length + site.files.length;
  log.info(
    `  <gray>${total} files generated in ${duration.toFixed(2)} seconds</gray>`,
  );
}
