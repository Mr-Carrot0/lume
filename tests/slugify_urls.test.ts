import { assertStrictEquals as equals } from "../deps/assert.ts";
import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import slugifyUrls from "../plugins/slugify_urls.ts";
import createSlugifier from "../core/slugifier.ts";
import unidecode from "npm:unidecode@1.1.0";
import jaconv from "npm:jaconv@1.0.4";

Deno.test("slugify_urls plugin", async (t) => {
  const site = getSite({
    src: "slugify_urls",
  });

  site.use(slugifyUrls());
  site.add([".css"]);
  site.add("_headers");

  await build(site);
  await assertSiteSnapshot(t, site);
});

Deno.test("slugify_urls plugin (static files)", async (t) => {
  const site = getSite({
    src: "slugify_urls",
  });

  site.use(slugifyUrls({
    extensions: "*",
  }));
  site.add([".css"]);
  site.add("_headers");

  await build(site);
  await assertSiteSnapshot(t, site);
});

Deno.test("slugify clean urls", () => {
  const slugify = createSlugifier();

  equals(slugify("Hello / World"), "hello/world");
  equals(slugify("hello_/_world"), "hello/world");
  equals(slugify("200,000*7"), "200-000-7");
  equals(slugify("img/Image 2 "), "img/image-2");
});

Deno.test("slugify replacement chars", () => {
  const slugify = createSlugifier();

  equals(slugify("Ð"), "d");
  equals(slugify("ð"), "d");
  equals(slugify("№"), "no");
});

Deno.test("slugify forbidden characters and words", () => {
  const slugify = createSlugifier({
    stopWords: ["or", "and", "the"],
  });

  equals(slugify(""), "");
  equals(slugify("the man or the woman"), "man-woman");
  equals(slugify("  forks   and   knives   "), "forks-knives");
  equals(slugify("went down the road"), "went-down-road");
});

Deno.test("slugify support unicode characters", () => {
  const slugify = createSlugifier({
    transliterate: {
      zh: unidecode,
      ja: jaconv.toHebon,
    },
  });

  equals(
    slugify("Lume 支持中文，中文标点。？、【】｛｝！￥（）", "zh"),
    "lume-zhi-chi-zhong-wen-zhong-wen-biao-dian-y",
  );

  equals(
    slugify("Lume はガリシア語で炎と言う意味でルメと発音する", "ja"),
    "lume-hadetoudetosuru",
  );
});
