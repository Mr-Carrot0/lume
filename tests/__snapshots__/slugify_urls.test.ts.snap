export const snapshot = {};

snapshot[`slugify_urls plugin 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      isPage: true,
      loader: [AsyncFunction: toml],
    },
    {
      engines: 1,
      ext: ".page.ts",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".page.js",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      engines: 0,
      ext: ".page.json",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".vto",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      ext: ".html",
    },
    {
      ext: ".css",
    },
  ],
  src: [
    "/",
    "/Chourizos ao viño.md",
    "/Page 1.md",
    "/UI Styles.css",
    "/_headers",
    "/generator.page.js",
    "/page-3.md",
    "/page-4.md",
  ],
}
`;

snapshot[`slugify_urls plugin 2`] = `
[
  {
    entry: "/_headers",
    flags: [],
    outputPath: "/_headers",
  },
  {
    entry: "/UI Styles.css",
    flags: [],
    outputPath: "/UI Styles.css",
  },
]
`;

snapshot[`slugify_urls plugin 3`] = `
[
  {
    content: "Foo",
    data: {
      basename: "generator",
      children: "Foo",
      content: "Foo",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/803nfon-sfodij/-/",
    },
    src: {
      ext: ".page.js",
      path: "/generator[1]",
      remote: undefined,
    },
  },
  {
    content: "<p>Page 2</p>
",
    data: {
      basename: "Chourizos ao viño",
      children: "<p>Page 2</p>
",
      content: "Page 2
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/chourizos-ao-vino/",
    },
    src: {
      ext: ".md",
      path: "/Chourizos ao viño",
      remote: undefined,
    },
  },
  {
    content: "Filloas con nocilla",
    data: {
      basename: "generator",
      children: "Filloas con nocilla",
      content: "Filloas con nocilla",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/filloas-con-nocilla/",
    },
    src: {
      ext: ".page.js",
      path: "/generator[0]",
      remote: undefined,
    },
  },
  {
    content: "<p>Page 1</p>
",
    data: {
      basename: "Page 1",
      children: "<p>Page 1</p>
",
      content: "Page 1
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/page-1/",
    },
    src: {
      ext: ".md",
      path: "/Page 1",
      remote: undefined,
    },
  },
  {
    content: "<p>Page 4</p>
",
    data: {
      basename: "page-4",
      children: "<p>Page 4</p>
",
      content: "Page 4
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/paxina-4.html",
    },
    src: {
      ext: ".md",
      path: "/page-4",
      remote: undefined,
    },
  },
  {
    content: "<p>Page 3</p>
",
    data: {
      basename: "page-3",
      children: "<p>Page 3</p>
",
      content: "Page 3
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/paxina-numero/tres/",
    },
    src: {
      ext: ".md",
      path: "/page-3",
      remote: undefined,
    },
  },
]
`;

snapshot[`slugify_urls plugin (static files) 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      isPage: true,
      loader: [AsyncFunction: toml],
    },
    {
      engines: 1,
      ext: ".page.ts",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".page.js",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      engines: 0,
      ext: ".page.json",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".vto",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      ext: ".css",
    },
  ],
  src: [
    "/",
    "/Chourizos ao viño.md",
    "/Page 1.md",
    "/UI Styles.css",
    "/_headers",
    "/generator.page.js",
    "/page-3.md",
    "/page-4.md",
  ],
}
`;

snapshot[`slugify_urls plugin (static files) 2`] = `
[
  {
    entry: "/_headers",
    flags: [],
    outputPath: "/headers",
  },
  {
    entry: "/UI Styles.css",
    flags: [],
    outputPath: "/ui-styles.css",
  },
]
`;

snapshot[`slugify_urls plugin (static files) 3`] = `
[
  {
    content: "Foo",
    data: {
      basename: "generator",
      children: "Foo",
      content: "Foo",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/803nfon-sfodij/-/",
    },
    src: {
      ext: ".page.js",
      path: "/generator[1]",
      remote: undefined,
    },
  },
  {
    content: "<p>Page 2</p>
",
    data: {
      basename: "Chourizos ao viño",
      children: "<p>Page 2</p>
",
      content: "Page 2
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/chourizos-ao-vino/",
    },
    src: {
      ext: ".md",
      path: "/Chourizos ao viño",
      remote: undefined,
    },
  },
  {
    content: "Filloas con nocilla",
    data: {
      basename: "generator",
      children: "Filloas con nocilla",
      content: "Filloas con nocilla",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/filloas-con-nocilla/",
    },
    src: {
      ext: ".page.js",
      path: "/generator[0]",
      remote: undefined,
    },
  },
  {
    content: "<p>Page 1</p>
",
    data: {
      basename: "Page 1",
      children: "<p>Page 1</p>
",
      content: "Page 1
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/page-1/",
    },
    src: {
      ext: ".md",
      path: "/Page 1",
      remote: undefined,
    },
  },
  {
    content: "<p>Page 4</p>
",
    data: {
      basename: "page-4",
      children: "<p>Page 4</p>
",
      content: "Page 4
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/paxina-4.html",
    },
    src: {
      ext: ".md",
      path: "/page-4",
      remote: undefined,
    },
  },
  {
    content: "<p>Page 3</p>
",
    data: {
      basename: "page-3",
      children: "<p>Page 3</p>
",
      content: "Page 3
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/paxina-numero/tres/",
    },
    src: {
      ext: ".md",
      path: "/page-3",
      remote: undefined,
    },
  },
]
`;
